File: app/src/main/java/com/innov8/memeit/Activities/AuthActivity.java
package com.innov8.memeit.Activities;

import android.app.Activity;
import android.content.Intent;
import android.os.Build;
import android.os.Bundle;
import android.util.DisplayMetrics;

import com.google.android.material.snackbar.Snackbar;
import com.innov8.memeit.CustomClasses.CustomMethods;
import com.innov8.memeit.CustomClasses.ScrollingImageDrawable;
import com.innov8.memeit.Fragments.LoginFragment;
import com.innov8.memeit.Fragments.SetupFragment;
import com.innov8.memeit.Fragments.SignUpFragment;
import com.innov8.memeit.KUtilsKt;
import com.innov8.memeit.R;
import com.memeit.backend.MemeItClient;
import com.theartofdev.edmodo.cropper.CropImage;

import androidx.appcompat.app.AppCompatActivity;
import androidx.coordinatorlayout.widget.CoordinatorLayout;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentManager;

public class AuthActivity extends AppCompatActivity {
    public static final String STARTING_FRAGMENT_PARAM = "frag";
    public static final int FRAGMENT_LOGIN = 0;
    public static final int FRAGMENT_SIGNUP = 1;
    public static final int FRAGMENT_SETUP = 2;

    public SignUpFragment signUpFragment;
    public LoginFragment loginFragment;
    public SetupFragment setupFragment;
    CoordinatorLayout authCoordinate;
    private int currentFrag;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        CustomMethods.makeWindowSeamless(this);
        setContentView(R.layout.activity_auth);
        authCoordinate=findViewById(R.id.auth_coordinate);

        authCoordinate.setBackground(new ScrollingImageDrawable(this,R.drawable.many_pics_optimized,
                KUtilsKt.getScreenWidth(),
                KUtilsKt.getScreenHeight()));
        initFragments();
        if (getIntent() != null) {
            int i = getIntent().getIntExtra(STARTING_FRAGMENT_PARAM, FRAGMENT_LOGIN);
            setCurrentFragment(i);
        } else {
            setCurrentFragment(FRAGMENT_LOGIN);
        }

    }

    private Fragment getFragmentForInt(int i) {
        switch (i) {
            case FRAGMENT_LOGIN:
                return loginFragment;
            case FRAGMENT_SIGNUP:
                return signUpFragment;
            case FRAGMENT_SETUP:
                return setupFragment;
            default:
                throw new IllegalArgumentException("Index should have been either 0,1 or 2!");
        }
    }

    private void initFragments() {
        signUpFragment = new SignUpFragment();
        loginFragment = new LoginFragment();
        setupFragment = new SetupFragment();
    }

    public void setCurrentFragment(int i) {
        currentFrag = i;
        FragmentManager fm = getSupportFragmentManager();
        fm.beginTransaction()
                .replace(R.id.holder, getFragmentForInt(i))
                .commit();
    }

    /*This method returns the height of the navigation buttons*/
    public static int getSoftButtonsBarHeight(Activity a) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR1) {
            DisplayMetrics metrics = new DisplayMetrics();
            a.getWindowManager().getDefaultDisplay().getMetrics(metrics);
            int usableHeight = metrics.heightPixels;
            a.getWindowManager().getDefaultDisplay().getRealMetrics(metrics);
            int realHeight = metrics.heightPixels;
            return realHeight > usableHeight ? realHeight - usableHeight : 0;
        }

        return 0;
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (requestCode == MemeItClient.Auth.GOOGLE_SIGN_IN_REQUEST_CODE) {
            getFragmentForInt(currentFrag).onActivityResult(requestCode, resultCode, data);
        }
        else if (requestCode == CropImage.CROP_IMAGE_ACTIVITY_REQUEST_CODE) {
            getFragmentForInt(currentFrag).onActivityResult(requestCode, resultCode, data);
        }
    }

    public void showError(String error){
        Snackbar sb=Snackbar.make(authCoordinate,error,Snackbar.LENGTH_LONG);
        sb.show();
    }

}

File: app/src/main/java/com/innov8/memeit/activities/AuthActivity.kt
package com.innov8.memeit.activities

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.text.method.LinkMovementMethod
import android.view.KeyEvent
import android.view.inputmethod.InputMethodManager
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import androidx.constraintlayout.widget.ConstraintSet
import androidx.transition.TransitionManager
import com.facebook.CallbackManager
import com.google.android.material.snackbar.Snackbar
import com.innov8.memeit.activities.authModes.*
import com.innov8.memeit.utils.CustomMethods
import com.innov8.memeit.utils.ScrollingImageDrawable
import com.innov8.memeit.R
import com.innov8.memeit.utils.*
import com.innov8.memeit.activities.authModes.*
import kotlinx.android.synthetic.main.activity_auth.*

class AuthActivity : AppCompatActivity() {
    private var mLoading: Boolean = false
    fun setLoading(loading: Boolean, then: () -> Unit = {}) {
        mLoading = loading
        if (loading) {
            action_btn.startAnimation()
        } else {
            action_btn.revertAnimation(then)
        }
    }

    val introMode = IntroAuthMode(this)
    val signupMode = SignUpAuthMode(this)
    val googleSignupMode = GoogleSignupMode(this)
    val facebookSignupMode = FacebookSignupMode(this)
    val loginMode = LogInAuthMode(this)
    val personalizeMode = PersonalizeAuthMode(this)
    var currentMode: AuthMode = introMode
    val onError = { message: String ->
        setLoading(false)
        showError(message)
    }
    val callbackManager by lazy { CallbackManager.Factory.create() }

    private lateinit var constraintSet: ConstraintSet
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        CustomMethods.makeWindowSeamless(this)
        setContentView(R.layout.activity_auth)
        auth_root.background = ScrollingImageDrawable(this, R.drawable.many_memes,
                screenWidth,
                screenHeight)
        constraintSet = ConstraintSet()
        constraintSet.clone(auth_root)
        introMode.init()
        signupMode.init()
        loginMode.init()
        personalizeMode.init()

        dots_indicator.setViewPager(intro_pager)
        action_btn.setOnClickListener {
            execOnAction()
        }
        auth_question.movementMethod = LinkMovementMethod()

        signup_facebook.setOnClickListener {
            if (mLoading) {
                showError("Please wait a bit")
                return@setOnClickListener
            }
            currentMode.onFacebook()
        }
        signup_google.setOnClickListener {
            if (mLoading) {
                showError("Please wait a bit")
                return@setOnClickListener
            }
            currentMode.onGoogle()
        }
        val function: (TextView?, Int?, KeyEvent?) -> Boolean = { v, _, _ ->
            if (v == currentMode.getLastField()?.editText) {
                val mgr = this.getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
                mgr.hideSoftInputFromWindow(v?.windowToken, 0)
                execOnAction()
                true
            } else false
        }
        username_field.editText!!.setOnEditorActionListener(function)
        email_field.editText!!.setOnEditorActionListener(function)
        password_field.editText!!.setOnEditorActionListener(function)
        confirm_password_field.editText!!.setOnEditorActionListener(function)
        val modeInt = intent?.getIntExtra(STARTING_MODE_PARAM, MODE_INTRO) ?: MODE_INTRO
        setMode(getModeForInt(modeInt), withAnimation = false)
    }

    fun onGoto() {
        if (mLoading) {
            showError("Please wait a bit")
            return
        }
        currentMode.onGoto()
    }

    private fun execOnAction() {
        if (mLoading) {
            showError("Please wait a bit")
            return
        }
        if (currentMode.verifable()) {
            if (currentMode.onVerify()) currentMode.onAction()
        } else currentMode.onAction()
    }

    private fun getModeForInt(modeInt: Int) = when (modeInt) {
        MODE_LOGIN -> loginMode
        MODE_SIGNUP -> signupMode
        MODE_PERSONALIZE -> personalizeMode
        else -> introMode
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        currentMode.onActivityResult(requestCode, resultCode, data)
    }

    override fun onBackPressed() {
        setLoading(false)
        if (!currentMode.onBackPressed()) {
            super.onBackPressed()
        }
    }

    fun showError(error: String) {
        val sb = Snackbar.make(auth_root, error, Snackbar.LENGTH_LONG)
        sb.show()
    }

    companion object {
        const val STARTING_MODE_PARAM = "frag"
        const val MODE_INTRO = 0
        const val MODE_LOGIN = 1
        const val MODE_SIGNUP = 2
        const val MODE_PERSONALIZE = 3
    }

    fun setMode(mode: AuthMode, bundle: Bundle? = null, withAnimation: Boolean = true) {
        mode.applyContraint(constraintSet)
        if (withAnimation) TransitionManager.beginDelayedTransition(auth_root)
        constraintSet.applyTo(auth_root)
        mode.updateElements()
        if (bundle != null) mode.withBundle(bundle)
        currentMode = mode
    }


}

